### PR TITLE
refactor(explore): improve typing for Dnd controls

### PR DESCRIPTION
**Pull request by ktmud**
### SUMMARY

Improve typing for Explore Dnd controls. Some principles:

1. React component props always go with the component itself---unless props are shared between multiple components
2. Reuse [`ControlProps`](https://github.com/apache/superset/blob/6e8ff55bfca4ea8b0004bcf20577ba394ab3ff9f/superset-frontend/src/explore/components/Control.tsx#L29) to get a more complete shared props propagated from parent.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by ktmud** [comment]
[HOLD] the DnD Metrics control has a bug where adding a new metric will wipe out existing metrics. The typing for the MetricControl is suboptimal, which is probably why the bug exists in the first place...

@kgabryje I'll try to fix this in a separate PR unless you are already on it.

File: superset-frontend/src/explore/components/DatasourcePanel/index.tsx [hunk 1]
```diff
@@ -193,61 +194,64 @@ export default function DataSourcePanel({
...
-          ],
-          keepDiacritics: true,
-          baseSort: (a, b) =>
-            Number(b.item.is_certified) - Number(a.item.is_certified) ||
-            String(a.rankedValue).localeCompare(b.rankedValue),
-        }),
-      });
-    }, FAST_DEBOUNCE),
-    [columns, metrics],
+  const search = debounce(
```
Comment (kgabryje) [inline superset-frontend/src/explore/components/DatasourcePanel/index.tsx:197]: Is there a reason why you changed the order here? According to https://dmitripavlutin.com/react-throttle-debounce/ `useCallback` should go first (although it seems that what we had here wasn't fully correct as we should wrap debounce function in useMemo...)

**Comment by kgabryje** [comment]
> [HOLD] The UX for overriding single select MetricControl is quite confusing. While Drag & Drop, There is no clear indication that a control will only accept one metric.

We indicate that a control accepts only 1 value by not displaying ghost button when a value is added. However, I do agree that we don't communicate clearly that dropping another value will replace the first one...

File: superset-frontend/src/explore/components/DatasourcePanel/index.tsx (hunk 1 above)
Comment (ktmud) [reply superset-frontend/src/explore/components/DatasourcePanel/index.tsx:197]: ESLint react-hook plugin will complain not being able to detect exhaustive deps and suggests making the callback inline.

Comment (kgabryje) [reply superset-frontend/src/explore/components/DatasourcePanel/index.tsx:197]: I've just verified - simply replacing useCallback with useMemo solves the eslint problem
```
const search = useMemo(
    () =>
      debounce((value: string) => {
...
```

**Comment by junlincc** [comment]
thanks this is awesome! @ktmud do you mind splitting this pr into 2 one for improving typing one for logging?

File: superset-frontend/src/explore/components/DatasourcePanel/index.tsx (hunk 1 above)
Comment (ktmud) [reply superset-frontend/src/explore/components/DatasourcePanel/index.tsx:197]: Cool. That should work!

File: superset-frontend/src/explore/components/controls/DndColumnSelectControl/DndMetricSelect.tsx [hunk 2]
```diff
@@ -104,6 +108,12 @@ const columnsContainAllMetrics = (
   );
 };
 
+export type DndMetricSelectProps = DndControlProps<ValueType> & {
+  savedMetrics: savedMetricType[];
+  columns: ColumnMeta[];
+  datasourceType?: DatasourceType;
+};
+
 export const DndMetricSelect = (props: any) => {
```
Comment (ktmud) [inline superset-frontend/src/explore/components/controls/DndColumnSelectControl/DndMetricSelect.tsx:117]: TODO: get rid of this `any`. Currently `DndMetricSelectProps` will conflict with too many things within this function. This file needs a bigger overhaul.

**Comment by ktmud** [comment]
@kgabryje do you mind taking another look?

**Review by kgabryje — APPROVED** [review]
That's a HUGE improvement, thanks!

**Pull request merged**